Dns trainscore restructure

NathanSegerlind opened the pull request:
This separates the logic of the DNS pipeline into three distinct phases:
1. training a model
2. scoring a model
3. adding "side information" expected by OA to the output 

steps are independent save explicitly passed parameters from top-level analysis driver

It's more of "pull it and explore the new code in you IDE" kind of PR than a "study it with diff tools" kind of PR

NathanSegerlind committed:
dns_trainscore_restructure

rename of DNSSuspiciousConnectsAnalysis and comments

NathanSegerlind committed:
Merge branch 'dev' into dns_trainscore_restructure

NathanSegerlind committed:
dns_trainscore_restructure

moving feature creation from word creation into top level analysis

NathanSegerlind committed:
dns_trainscore_restructure

formatting

NathanSegerlind committed:
Merge branch 'dev' into dns_trainscore_restructure

NathanSegerlind committed:
Merge branch 'dev' into dns_trainscore_restructure

NathanSegerlind committed:
dns_trainscore_restructure

reconfigured training and scoring to use shared word
creation logic so that:

incoming data schema is always the same

previously, word creation relied on pre-existence of derived fields that are not present in incoming data

NathanSegerlind committed:
dns_trainscore_restructure

it runs and produces plausible output

NathanSegerlind committed:
dns_trainscore_restructure

moved calculation of the "side information"  for the OA layer outside the model proper

NathanSegerlind committed:
dns_trainscore_restructure

schema management, moving of domain logic to the org.opennetworkinsight.utilities.DomainProcessor object

NathanSegerlind committed:
Merge branch 'dev' into dns_trainscore_restructure

- had to delete no longer in-use DNSPostLDA, DNSPreLDA, DNSSuspiciousConnects files
- replaced slf4j logger with apache logger
- threw in some scaladoc while I was at it

# Conflicts:
#	src/main/scala/org/opennetworkinsight/dns/DNSPostLDA.scala
#	src/main/scala/org/opennetworkinsight/dns/DNSPreLDA.scala
#	src/main/scala/org/opennetworkinsight/dns/DNSSuspiciousConnects.scala
#	src/main/scala/org/opennetworkinsight/dns/DNSWordCreation.scala

NathanSegerlind committed:
dns_trainscore_restructure

NathanSegerlind commented:
This version is slightly faster than the old version because the "side info" columns get added only to the (usually much smaller) collection of events that pass the filter and are considered suspicious

rabarona commented on src/main/scala/org/opennetworkinsight/dns/DNSSuspiciousConnectsAnalysis.scala:
```diff
@@ -0,0 +1,93 @@
+  def run(config: SuspiciousConnectsConfig, sparkContext: SparkContext, sqlContext: SQLContext, logger: Logger) = {
+    logger.info("Starting DNS suspicious connects analysis.")
+
+    val topicCount = 20
```
Is this the same as mpiTopicCount in config: SuspiciousConnectsConfig? If so, it would just need to be converted to an integer.

NathanSegerlind replied:
yeah, should be all coming from the same place

I'll do a quick smoke test and then push my changes

NathanSegerlind commented on src/main/scala/org/opennetworkinsight/dns/DNSPostLDA.scala:
```diff
@@ -1,94 +0,0 @@
-
```
this stuff went mostly into the "score" method of the model

NathanSegerlind commented on src/main/scala/org/opennetworkinsight/dns/DNSPreLDA.scala:
```diff
@@ -1,136 +0,0 @@
-package org.opennetworkinsight.dns
```
this stuff went mostly into the "train" part of the model

NathanSegerlind commented on src/main/scala/org/opennetworkinsight/dns/DNSSuspiciousConnects.scala:
```diff
@@ -1,31 +0,0 @@
-package org.opennetworkinsight.dns
-
```
renamed to DNSSuspiciousConnectsAnalysis

NathanSegerlind commented on src/main/scala/org/opennetworkinsight/dns/DNSSuspiciousConnectsAnalysis.scala:
```diff
@@ -0,0 +1,93 @@
+  val inColumns = inSchema.fieldNames.map(col)
+
+
+  assert(ModelSchema.fields.forall(inSchema.fields.contains(_)))
```
should add a similar validation for the side field schema and outschema...

the idea is to catch a potential schema mismatch created by a customizing user as early as possible...

the biggest drawback with using dataframes is that schema checks occur only at runtime, not during typechecking... but I think that the easy management of columns outweighs that

still, we should make it so that any mucking about with the input/model/side/output schemas is validated before things get going

NathanSegerlind commented on src/main/scala/org/opennetworkinsight/dns/DNSSuspiciousConnectsAnalysis.scala:
```diff
@@ -0,0 +1,93 @@
+    val mostSusipiciousDF: DataFrame = filteredDF.orderBy(Score).limit(config.maxResults)
+
+    // add the "side information" expected by the OA layer
+    val sideInformationGenerator = new DNSSideInformation(model)
```
honestly, I don't like doing this at all and we should revisit why we are providing this information at all (why can't the OA compute the values it needs?) this is an ugly complication, especially since it shares some code that would otherwise be solely the concern of the model

rabarona replied:
I think the main reason is because of the time it would take to compare the compose key against all the data in HDFS for that day or time frame. Anyway, it represents processing time and I think it can be either done here or in OA; I share the feeling, OA as an ETL process should be in charge. I can start checking for an option, maybe pyspark to query the required data based on the records that are planned to go to the UI.

NathanSegerlind replied:
there won't be any changes in this PR, but we should get the ball rolling for future discussion

in theory, I believe that each piece of code should have a single responsibility 
in practice, I recognize that sometimes it's way cheaper to spin up one batch job that does it all than to do a bunch of one-off jobs

NathanSegerlind committed:
Merge branch 'dns_trainscore_restructure' of github.com:NathanSegerlind/oni-ml into dns_trainscore_restructure

# Conflicts:
#	src/main/scala/org/opennetworkinsight/dns/DNSWordCreation.scala

NathanSegerlind commented on src/main/scala/org/opennetworkinsight/dns/DNSWordCreation.scala:
```diff
@@ -1,239 +1,95 @@
 package org.opennetworkinsight.dns
 
```
proxy uses some of this same logic so it was moved to a shared DomainProcessor utitility

NathanSegerlind commented on src/main/scala/org/opennetworkinsight/dns/DNSWordCreation.scala:
```diff
@@ -1,239 +1,95 @@
+  * @param numberPeriodsCuts
+  * @param topDomainsBC
+  */
+class DNSWordCreation(frameLengthCuts: Array[Double],
```
do these parameters look familiar?

they are basically exactly the state in a DNSSuspiciousConnectsModel....

if it weren't for the OA side information requiring the "words" we could just make this a method in the DNSSuspiciousConnectsModel  (or a closely tied Serializable class)

NathanSegerlind commented on src/main/scala/org/opennetworkinsight/dns/model/DNSScoreFunction.scala:
```diff
@@ -0,0 +1,57 @@
+  * @param topDomainsBC
+  */
+class DNSScoreFunction(frameLengthCuts: Array[Double],
+                       timeCuts: Array[Double],
```
I don't like passing in almost all the state from the DNSSuspiciousConnectsModel,  but I don't want to have to Serialize that thing....

NathanSegerlind commented on src/test/scala/org/opennetworkinsight/DNSWordCreationTest.scala:
```diff
@@ -1,93 +1,21 @@
+import org.opennetworkinsight.utilities.{CountryCodes, Entropy, TopDomains}
 import org.scalatest.Matchers
 
-class DNSWordCreationTest extends TestingSparkContextFlatSpec with Matchers{
```
moved to the domain processor test suite

NathanSegerlind commented on src/test/scala/org/opennetworkinsight/utilities/DomainProcessorTest.scala:
```diff
@@ -1,68 +1,71 @@
 package org.opennetworkinsight.utilities
```
crap

I dropped these tests in some merge noise

I will add them back and repush

NathanSegerlind commented on src/main/scala/org/opennetworkinsight/utilities/DomainProcessor.scala:
```diff
@@ -2,12 +2,12 @@ package org.opennetworkinsight.utilities
 
```
hmmm

maybe we should just pull in a third-party open source library for this stuff

https://github.com/NET-A-PORTER/scala-uri

is Apache 2 licensed

but who knows how long it will be supported, etc.

NathanSegerlind committed:
dns_trainscore_restructure

cleanup

rabarona commented on src/main/scala/org/opennetworkinsight/proxy/ProxySuspiciousConnectsModel.scala:
```diff
@@ -172,7 +172,7 @@ object ProxySuspiciousConnectsModel {
+    val topDomains: Broadcast[Set[String]] = sc.broadcast(TopDomains.TopDomains)
 
     val agentToCountBC = sc.broadcast(agentToCount)
     val udfWordCreation = ProxyWordCreation.udfWordCreation(topDomains, agentToCountBC, timeCuts, entropyCuts, agentCuts)
```
@NathanSegerlind I don't want to be too picky but seems like a new convention was defined during this restructure. Looks like udf functions are called like `functionNameUDF` but in this line, I can still see the old fashion naming.

rabarona reviewed (changes requested):
Maybe this is why you wasn't viewing any of my comments. I needed to submit a review.

NathanSegerlind commented on src/main/scala/org/opennetworkinsight/dns/DNSWordCreation.scala:
```diff
@@ -1,10 +1,100 @@
 package org.opennetworkinsight.dns
 
+<<<<<<< HEAD
```
these are not in the version I just pushed, I will see if I can force the file to be updated

NathanSegerlind committed:
dns_trainscore_restructure

code review comments:  simplification of command line options for suspicous connects analysis

NathanSegerlind commented:
regarding topicCount in  proxy:

that should be taken care of in the task that Julia is working on
